src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,377 @@
+            // only be an Insert or Delete, respectively.
+
+            using var _ = ArrayBuilder<LSP.SemanticTokensEdit>.GetInstance(out var semanticTokensEdits);
+
+            // We sort the edit kinds by index since we need to know what kind of edits surround
+            // a given index in order to potentially combine them into one edit.
+            var editIndexes = indexToEditKinds.Keys.ToArray();
+            Array.Sort(editIndexes);
+
+            for (var i = 0; i < editIndexes.Length; i++)
+            {
+                var initialEditIndex = editIndexes[i];
+                var initialEditKind = indexToEditKinds[initialEditIndex];
+
+                if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert) && initialEditKind.HasFlag(SemanticTokenEditKind.Delete))
+                {
+                    i = AddInsertionDeletionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else if (initialEditKind.HasFlag(SemanticTokenEditKind.Insert))
+                {
+                    i = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+                else
+                {
+                    i = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndexes, i, initialEditIndex);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            static int AddInsertionDeletionEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndexes,
+                int i,
+                int initialEditIndex)
+            {
+                var deleteCount = 5;
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                tokensToInsert.AddRange(newGroupedSemanticTokens[initialEditIndex].ToArray());
+
+                // An "update" (i.e. a dual insertion/deletion) can be combined with an update,
+                // deletion, or insertion that directly follows it. If combined with an insertion
+                // or deletion, only that type is allowed in the combined edit afterwards.
+                var editKind = SemanticTokenEditKind.None;
+
+                // To continue combining edits, we need to ensure:
+                // 1) There is an edit following the current edit
+                // 2) The current edit and next edit involve tokens that are located right next to
+                // each other in the file.
+                while (i + 1 < editIndexes.Length && editIndexes[i + 1] == editIndexes[i] + 1)
+                {
+                    var currentEditKind = indexToEditKinds[editIndexes[i + 1]];
```

Comment:
i woudl call this `nextEditKind`, because you're literally looking at the *next* thing.  The 'curent' is the one you already had.